design/0003-byop.md

```diff
@@ -0,0 +1,198 @@
+# Bring your own Prometheus
+
+- @pothulapati
+- Start Date: 2020-04-15
+- Target Date: 
+- RFC PR: [linkerd/rfc#16](https://github.com/linkerd/rfc/pull/16)
+- Linkerd Issue: [linkerd/linkerd2#3590](https://github.com/linkerd/linkerd2/issues/3590)
+- Reviewers: @grampelberg @alpeb @siggy
+
+## Summary
+
+[summary]: #summary
+
+This RFC aims to move prometheus as an add-on i.e not mandated but advised.
+This includes moving the linkerd-prometheus manifests as a sub-chart and
+allowing more configuration (more on this later), along with documentation
+and tooling to use Linkerd with an existing prometheus installation.
+
+## Problem Statement (Step 1)
+
+[problem-statement]: #problem-statement
+
+Lots of users already have a exisiting prometheus installation in their
+cluster and don't like managing another prometheus instance especially
+because they could be resource intensive and painful. They usually
+have a prometheus installation paired up with a long term monitroing
+solutions like Cortex,InfluxDB, or third party solutions like Grafana
+Cloud, etc. In those cases prometheus is just a temporary component, that
+keeps pushing metrics into their long term solution, making users want to
```

Comment:
I'm not sure I really understand what users don't like about the existence of more than one Prometheus in their cluster.  What is the actual downside to having Linkerd-prometheus as a temporary storage for metrics before they are pulled into long-term storage via federation or remote-writes?  Does the linkerd-prometheus have a high resource cost?  Can we quantify this?  Is there some greater operational burden to having the linkerd-prometheus exist?  what is that burden?

Reply:
Yep, `linkerd-prometheus` isn't any different from the normal installation, and dosen't also get any specific tweaks for Linkerd. I definitely think users should be using the `linkerd-prometheus` as it dosen't do anything different and the operational burden shouldn't matter that much if there is already a Prometheus and long-term solution. We will be suggesting the same in the documentation.

On the resource cost,
We have the following in `ha` mode:
```
prometheusResources:
  cpu:
    limit: "4"
    request: 300m
  memory:
    limit: 8192Mi
    request: 300Mi
```

This isn't a lot but we have definitely seen cases where prometheus was being OOM, etc. I don't think this will be any better in the BYOP case, and the problem is further amplified. 

I think, It's just one of those obv questions that users stumble upon initially "Can I use my existing prometheus?" but it would later make sense as they start using, to have separate one for Linkerd. 

IMO the main value of this rfc is separating out prometheus from the contorl-plane components in charts, configuration and having a life-cycle of its own for updating configuration, etc. BYOP is just another benefit that we are getting out of this. WDYT?